### PR TITLE
Onboarding and Authentication Flow

### PR DESCRIPTION
**Pull request by yadunut**
This PR adds the flow for authentication and onboarding.

Some things to consider. 

`UserRepositoryInterface` handles all API / Persistence action for User Related Data. I'm calling it User for now, but we can refactor to student or something else. (Called User as there is both public and students using it).

`AuthenticationBloc` Handles the starting of the app. It shows the splash screen, and once the authentication module is initialized / waits a predefined amount of time, it goes to the next screen. 
It checks whether the user is a public user / has already logged in. If they have, it will jump straight to the HomeScreen. If they havent, it'll go to the Welcome screen and start the onboarding Process. 

I don't do checks whether the token is valid or invalid but I don't think that would be necessary.

The design aspect of this will be added in another PR. 
Actual connections with graphQL and other backends will also be added in a future PR.

### CODE REVIEW

**Comment by yadunut** [comment]
@ravernkoh @seanlim Pls look through, if no issues I'll be merging tonight.

File: mobile-app/lib/screens/onboarding/login_screen.dart [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            authenticationBloc: BlocProvider.of<AuthenticationBloc>(context),
+            userRepository: userRepository),
+        child: LoginForm(),
+      ),
+    );
+  }
+}
+
+class LoginForm extends StatefulWidget {
+  LoginForm({Key key}) : super(key: key);
```
Comment (ravernkoh) [inline mobile-app/lib/screens/onboarding/login_screen.dart]: What's this for?

File: mobile-app/lib/main.dart [hunk 2]
```diff
@@ -1,13 +1,76 @@
...
-        title: 'Water Learning Programme', home: StationsScreen());
+      title: 'Water Learning Programme',
+      home: BlocBuilder<AuthenticationBloc, AuthenticationState>(
+        builder: (context, state) {
+          var animatedSwitch = (Widget widget) {
+            return AnimatedSwitcher(
+              child: widget,
+              duration: Duration(seconds: 1),
+            );
+          };
```
Comment (ravernkoh) [inline mobile-app/lib/main.dart]: Why not declare this as a top-level function?

File: mobile-app/lib/constants.dart [hunk 3]
```diff
@@ -0,0 +1,3 @@
+import 'package:flutter/material.dart';
+
+const blue = Color(0xFF005DE8);
```
Comment (ravernkoh) [inline mobile-app/lib/constants.dart]: I think Flutter has built-in theming.

File: mobile-app/lib/bloc/authentication.dart [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  LoggedIn({@required this.token});
+}
+
+class LoggedOut extends AuthenticationEvent {}
+
+class IsPublic extends AuthenticationEvent {}
+
+class IsStudent extends AuthenticationEvent {}
+
+class AuthenticationBloc
```
Comment (ravernkoh) [inline mobile-app/lib/bloc/authentication.dart]: Why not call it `OnboardingBloc`?

File: mobile-app/lib/screens/onboarding/login_screen.dart (hunk 1 above)
Comment (yadunut) [reply mobile-app/lib/screens/onboarding/login_screen.dart]: Oh oops I was using a snippet. Autogenerated...

File: mobile-app/lib/main.dart (hunk 2 above)
Comment (yadunut) [reply mobile-app/lib/main.dart]: Hmmm I wasn't going to use it in other places and having an anonymous function seemed better. I'll switch it to a function call but I don't think it needs to be top level. Or are there any advantages to having a top level function?
```dart
_animatedSwitch(Widget widget) {
  return AnimatedSwitcher(
    child: widget,
    duration: Duration(seconds: 1),
   );
};

File: mobile-app/lib/constants.dart (hunk 3 above)
Comment (yadunut) [reply mobile-app/lib/constants.dart]: I'll be addressing this in the future PR with updating the designs.

File: mobile-app/lib/main.dart (hunk 2 above)
Comment (ravernkoh) [reply mobile-app/lib/main.dart]: Not much. Just couldn't find the `animatedSwitch` function when scrolling through the file. Maybe `buildAnimatedSwitcher` would be better.

File: mobile-app/lib/bloc/authentication.dart (hunk 4 above)
Comment (yadunut) [reply mobile-app/lib/bloc/authentication.dart]: I'll rename this to `OnboardingBloc` but ill leave the rest as AuthenticationEvent and AuthenticationState as those make more sense.

Comment (ravernkoh) [reply mobile-app/lib/bloc/authentication.dart]: I disagree. Keeping things consistent is more important. Choose one.

**Pull request merged**